### ISSUE TITLE
hh_server child processes crash since 3.19.x

### ISSUE THREAD

**Issue by RyanGordon** [facebook/hhvm#7887]
### HHVM Version

HHVM Version 3.20.2 (ad133f299ece0e45db8536473a8ef77cc4612ae3). This problem first appeared in version 3.19.x. It does not happen on 3.18.1.

### Standalone code, or other way to reproduce the problem

Right now running it against our (private) code base causes the crash. I think it may have a problem when it reaches one of our files called xhprof.php. Haven't been able to isolate anything yet though.

I would love any advice if there is a way to get more information out of the crashed child process / debug better. The exit error code of 16 appears to be opaque. I can't seem to find any correlated error message, other than it may be a UNIX exit code?

```
/usr/local/bin/hh_server --debug /path/to/code
```

### Expected result

hh_server doesn't crash and runs normally.

### Actual result

```
/usr/local/bin/hh_server --debug /path/to/code
[2017-06-16 00:29:15] Loading config exception: Sys_error("/etc/hh.conf: No such file or directory")
[2017-06-16 00:29:15] Could not load config at /etc/hh.conf, using defaults
[2017-06-16 00:29:15] Starting first server
[2017-06-16 00:29:15] About to spawn typechecker daemon. Logs will go here.
[2017-06-16 00:29:15] Just started typechecker server with pid: 27394.
[2017-06-16 00:29:15] Loading config exception: Sys_error("/etc/hh.conf: No such file or directory")
[2017-06-16 00:29:15] Could not load config at /etc/hh.conf, using defaults
[2017-06-16 00:29:15] Version: ad133f299ece0e45db8536473a8ef77cc4612ae3 Thu Jun  8 14:59:12 2017
[2017-06-16 00:29:15] Hostname: dev-machine
[2017-06-16 00:29:15] Loading config exception: Sys_error("/etc/hh.conf: No such file or directory")
[2017-06-16 00:29:15] Could not load config at /etc/hh.conf, using defaults
[2017-06-16 00:29:15] No debug port attached
[2017-06-16 00:29:15] Initializing Server (This might take some time)
[2017-06-16 00:29:15] Init id: TTSL/fba9a
[2017-06-16 00:29:15] Indexing: 0.000007
[2017-06-16 00:29:17] Heap size: 20214080
[2017-06-16 00:29:17] Parsing: 2.270982
[2017-06-16 00:29:17] Updating deps: 0.041399
[2017-06-16 00:29:17] Heap size: 21499584
[2017-06-16 00:29:17] Naming: 0.088990
[2017-06-16 00:29:23] Heap size: 35841280
[2017-06-16 00:29:23] Type-check: 5.789967
Worker exited (code: 16)
Worker exited (code: 16)
Worker exited (code: 16)
Worker exited (code: 16)
Subprocess(27412): fail 16[2017-06-16 00:29:23] Worker.Worker_exited_abnormally(16)
Raised at file "procs/worker.ml", line 315, characters 14-42
Called from file "procs/worker.ml", line 362, characters 18-29
Re-raised at file "procs/worker.ml", line 372, characters 14-17
Called from file "procs/multiWorker.ml", line 77, characters 31-52
Called from file "list.ml", line 88, characters 24-34
Called from file "third-party/core/core_list.ml", line 192, characters 22-47
Called from file "procs/multiWorker.ml", line 87, characters 20-70
Called from file "server/serverInit.ml", line 710, characters 4-43
Called from file "server/serverInit.ml", line 753, characters 2-19
Called from file "server/serverMain.ml", line 447, characters 19-39
Called from file "server/serverMain.ml", line 36, characters 14-67
Called from file "server/serverMain.ml", line 540, characters 12-74
Called from file "server/serverUtils.ml", line 67, characters 6-10
[2017-06-16 00:29:24] typechecker exit with exit code 16
```

**Comment by mofarrell** [comment]
This seems to be a duplicate of https://github.com/hhvm/homebrew-hhvm/issues/86
Are you running on OS X as well?

**Comment by RyanGordon** [comment]
This is on CentOS 7, may have the root cause though

**Comment by RyanGordon** [comment]
I was able to track down that [error code 16](https://github.com/facebook/hhvm/blob/ae455db63643f66c245efcb3d35c16fe07eb710c/hphp/hack/src/utils/exit_status.ml#L99) means `HASH_TABLE_FULL`. It was not trivial to find this. It would be nice if it would display an actual stringified error in hh_client output

It turns out that if you set the `sharedmem_hash_table_pow` to 20 then we're back in business. It appears that sharedmem_hash_table_pow actually represents the value in kB rather than bytes and that whatever value you set it to, it needs to be able to virtually malloc that much even if it doesn't use most of it. For instance, on my development machine, hh_server has a virtual size of 21GB but really only uses a fraction of that in RSS.

It's good to know what the error code referred to after digging and what unit sharedmem_hash_table_pow represented after guesswork but none of this is documented and specifically sharedmem_hash_table_pow isn't even mentioned in [Typechecker: Setup](https://docs.hhvm.com/hack/typechecker/setup) along with a half dozen other options that are missing. The backtrace that was given was also essentially useless but maybe that's just a limitation of OCaml or the way the parent/child processes are setup in hh_server/client that make it hard.

In summary, this problem is "fixed", but it would be great to get some expanded documentation on hh_client and hh_server for the knowledge of other users.
